token/services/sql/tokens.go

```diff
@@ -0,0 +1,886 @@
+		found := false
+		for j := 0; j < len(inputs); j++ {
+			if inputs[j].Equal(tokID) {
+				ids[j] = id
+				tokens[j] = tok
+				logger.Debugf("set token at location [%s:%s]-[%d]", tok.Type, tok.Quantity, j)
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, nil, errors.Errorf("retrieved wrong token [%s]", id)
+		}
+
+		counter++
+	}
+	logger.Debugf("found [%d] tokens, expected [%d]", counter, len(inputs))
+	if err = rows.Err(); err != nil {
+		return nil, tokens, err
+	}
+	if counter == 0 {
+		return nil, nil, errors.Errorf("token not found for key [%s:%d]", inputs[0].TxId, inputs[0].Index)
+	}
+	if counter != len(inputs) {
+		for j, t := range tokens {
+			if t == nil {
+				return nil, nil, errors.Errorf("token not found for key [%s:%d]", inputs[j].TxId, inputs[j].Index)
+			}
+		}
+		panic("programming error: should not reach this point")
+	}
+	return ids, tokens, nil
+}
+
+// WhoDeletedTokens returns information about which transaction deleted the passed tokens.
+// The bool array is an indicator used to tell if the token at a given position has been deleted or not
+func (db *TokenDB) WhoDeletedTokens(ns string, inputs ...*token.ID) ([]string, []bool, error) {
+	logger.Debugf("search first over token table [%s]...", inputs)
+	who, deleted, err := db.whoDeleteTokens(ns, db.table.Tokens, inputs...)
+	if err != nil || len(who) != len(inputs) {
+		logger.Debugf("search then over auditor token table [%s]...", inputs)
+		return db.whoDeleteTokens(ns, db.table.AuditTokens, inputs...)
+	}
+	return who, deleted, err
+}
+
+func (db *TokenDB) whoDeleteTokens(ns string, table string, inputs ...*token.ID) ([]string, []bool, error) {
+	if len(inputs) == 0 {
+		return []string{}, []bool{}, nil
+	}
+	args := make([]interface{}, 0)
+	where := whereTokenIDs(&args, ns, inputs)
+
+	query := fmt.Sprintf("SELECT tx_id, idx, spent_by, is_deleted FROM %s WHERE %s", table, where)
+	logger.Debug(query, args)
+	rows, err := db.db.Query(query, args...)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer rows.Close()
+
+	spentBy := make([]string, len(inputs))
+	isSpent := make([]bool, len(inputs))
+
+	counter := 0
+	for rows.Next() {
+		var txid string
+		var idx uint64
+		var spBy string
+		var isSp bool
+		if err := rows.Scan(&txid, &idx, &spBy, &isSp); err != nil {
+			return spentBy, isSpent, err
+		}
+		// order is not necessarily the same, so we have to set it in a loop
+		for i, inp := range inputs {
+			if inp.TxId == txid && inp.Index == idx {
+				isSpent[i] = isSp
+				spentBy[i] = spBy
+				break // stop searching for this id but continue looping over rows
+			}
+		}
+		counter++
+	}
+	logger.Debugf("found [%d] records, expected [%d]", counter, len(inputs))
+	if err = rows.Err(); err != nil {
+		return nil, isSpent, err
+	}
+	if counter == 0 {
+		return nil, nil, errors.Errorf("token not found for key [%s:%d]", inputs[0].TxId, inputs[0].Index)
+	}
+	if counter != len(inputs) {
+		return nil, nil, errors.Errorf("record missings")
```

Comment:
Maybe we can be more descriptive, similar to the debug of line 716?

Reply:
👍